mitodl/mit-open
# Alert handler for percolate matches

### What are the relevant tickets?

Closes #783


### Description (What does it do?)
This PR adds the following:
1. A hook that piggy backs on document upserts and percolates queries when a new document is added to the database
2. An empty signal handler that receives a document along with queries that have been percolated for it -> the functionality that would go in the signal handler is still TBD

### How can this be tested?
1. Do a docker-compose down + up so we start with a clean slate (easiest way I have found to visibly test). 
2. Now that we have a fresh database and search index, create a super user and login
3. go to the search [subscribe view](http://od.odl.local:8063/api/v1/learning_resources_user_subscription/subscribe) and post some search subscriptions such as ```{"q":"test"}``` or ```{"q":"test", "topics":["science", "physics"]}``` - this view takes the same query parameters as the search page as post vars.
4. The signal handler that gets called when a document that matches percolates is [defined here](https://github.com/mitodl/mit-open/blob/shanbady/alert-handlers-for-percolate-matches/learning_resources_search/signals.py#L44) - add some more prominent logging there so you can catch the output in the celery logs. The handler receives a LearningResource as the "instance" parameter and percolated_queries as a kwarg which is a list of PercolateQuery instances.
5. get into the web docker instance and run one of the load data commands ex ```./manage.py backpopulate_micromasters_data```
6. Documents that are matched to your queries should be logged in the celery logs.


### Additional Context
The way this works is by using a [celery chain](https://docs.celeryq.dev/en/stable/userguide/canvas.html#chains) so that we percolate once the document has indexed.

I wanted to ensure that anytime a new document was added, we could percolate however one thing I noted locally was that LearningResources added through django-admin or any other way outside of an ETL loader does not get indexed or percolated. This is fine if we are assuming that the only way documents are added is via etl otherwise we need to find a common entrypoint. @mbertrand might have more context/advice around this 


### Checklist:
- [ ] end to end flow has been tested
